### PR TITLE
Prevent data leakage in ML preprocessing

### LINKED ISSUE #514: Bug: Data leakage — scaler/encoder fit_transform() called on full dataset before train/test split
## Bug Report

**Description**
In `Automated Financial Reporting with Deep Learning/app.py`, the `MinMaxScaler` is fitted on
the **entire dataset** (including what would be test data) before any train/test split. This
causes **data leakage** — the model indirectly sees test data statistics during training,
leading to artificially inflated accuracy metrics that won't hold in real-world use.

## Affected Files

| File | Line | Issue |
|------|------|-------|
| `Automated Financial Reporting with Deep Learning/app.py` | 16 | `scaler.fit_transform()` on entire df |
| `Engineering Career Role Prediction/app.py` | 32 | `le.fit_transform()` inside prediction function using training encoder on live input |

## Code (Buggy)
```python
scaler = MinMaxScaler()
data_scaled = scaler.fit_transform(df.drop('Date', axis=1))  # leaks test stats into training
```

## Fix
```python
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)   # fit ONLY on train
X_test_scaled = scaler.transform(X_test)          # transform test separately
```

## Severity
High — produces misleadingly optimistic model performance; this is a teaching repo so it
actively trains beginners with incorrect ML practices

### PR DESCRIPTION
**Pull request by akhila-sraj**
PR description:
This change fixes a data leakage issue in the training pipeline by ensuring preprocessing is fit only on the training split. Previously, scaling or encoding could be applied before the train/test split, allowing test-set statistics to influence training and inflate evaluation metrics.

What changed:

Split the dataset before fitting any preprocessing steps.
Fit the scaler or encoder on training data only.
Apply transform to validation or test data separately.
Keep inference-time preprocessing aligned with the training pipeline.
Why it matters:

Prevents optimistic accuracy estimates caused by leakage.
Makes evaluation metrics representative of real-world performance.
Reinforces correct ML practice for learners and contributors.

fixes #514

### CODE REVIEW

**Comment by akhila-sraj** [comment]
@SdSarthak This fixes issue Bug: Data leakage — scaler/encoder fit_transform() called on full dataset before train/test split #514 Could you please review it and merge the code
